zipline-loader/src/commonMain/kotlin/app/cash/zipline/loader/ZiplineLoader.kt

```diff
@@ -52,48 +125,145 @@ class ZiplineLoader(
 
   suspend fun load(
     zipline: Zipline,
     manifest: ZiplineManifest,
+  ) = receive(ZiplineLoadReceiver(zipline), manifest)
+
+  suspend fun download(
+    downloadDir: Path,
+    downloadFileSystem: FileSystem,
+    manifestUrl: String,
   ) {
-    ZiplineModuleLoader.createProduction(
-      dispatcher = dispatcher,
-      httpClient = httpClient,
-      concurrentDownloadsSemaphore = concurrentDownloadsSemaphore,
-      embeddedDir = embeddedDir,
-      embeddedFileSystem = embeddedFileSystem,
-      cache = cache,
-      zipline = zipline,
-    ).load(manifest)
+    val manifest = downloadZiplineManifest(manifestUrl)
+    download(downloadDir, downloadFileSystem, manifest)
   }
 
-  private suspend fun downloadZiplineManifest(manifestUrl: String): ZiplineManifest {
-    val manifestByteString = try {
-      concurrentDownloadsSemaphore.withPermit {
-        httpClient.download(manifestUrl)
+  suspend fun download(
+    downloadDir: Path,
+    downloadFileSystem: FileSystem,
+    manifest: ZiplineManifest,
+  ) {
+    writeManifestToDisk(downloadFileSystem, downloadDir, manifest)
+    receive(
+      receiver = FsSaveReceiver(downloadFileSystem, downloadDir),
+      manifest = manifest
+    )
+  }
+
+  private suspend fun receive(
+    receiver: Receiver,
+    manifest: ZiplineManifest,
+  ) {
+    coroutineScope {
+      val loads = manifest.modules.map {
+        ModuleJob(it.key, it.value, receiver)
+      }
+      for (load in loads) {
+        val loadJob = launch { load.run() }
+
+        val downstreams = loads.filter { load.id in it.module.dependsOnIds }
+        for (downstream in downstreams) {
+          downstream.upstreams += loadJob
+        }
       }
-    } catch (e: IOException) {
-      // If manifest fails to load over network, fallback to prebuilt in resources
-      val prebuiltManifestPath = embeddedDir / PREBUILT_MANIFEST_FILE_NAME
-      embeddedFileSystem.read(prebuiltManifestPath) {
-        readByteString()
+    }
+  }
+
+  private inner class ModuleJob(
+    val id: String,
+    val module: ZiplineModule,
+    val receiver: Receiver,
+  ) {
+    val upstreams = mutableListOf<Job>()
+
+    /**
+     * Fetch and receive ZiplineFile module
+     */
+    suspend fun run() {
+      val byteString = fetchers.fetch(concurrentDownloadsSemaphore, id, module.sha256, module.url)
+
+      upstreams.joinAll()
+      withContext(dispatcher) {
+        receiver.receive(byteString, id, module.sha256)
       }
     }
+  }
+
+  private suspend fun downloadZiplineManifest(
+    manifestUrl: String,
+  ): ZiplineManifest {
+    val manifestByteString = fetchers.fetch(
+      concurrentDownloadsSemaphore = concurrentDownloadsSemaphore,
+      id = PREBUILT_MANIFEST_FILE_NAME,
+      // Store manifest as the url SHA.
+      sha256 = manifestUrl.encodeUtf8().sha256(),
+      url = manifestUrl,
+      // Override fileName to support a static manifest location in embedded fs.
+      fileNameOverride = PREBUILT_MANIFEST_FILE_NAME,
```

Comment:
I think this needs to always come from the user. We want to support multiple embedded applications.